src/paradigma/heart_rate/heart_rate_estimation.py

```diff
@@ -0,0 +1,185 @@
+
+        # Correct for first and last epochs
+        if start_idx < 0:
+            start_idx = 0
+        if end_idx > len(ppg_prob):
+            end_idx = len(ppg_prob)
+
+        # Extract probabilities and labels for the current epoch
+        prob = ppg_prob[start_idx:end_idx+1]
+        label_imu = acc_label[start_idx:end_idx+1]
+
+        # Calculate mean probability and majority voting for labels
+        data_prob[i] = np.mean(prob)
+        data_label_imu[i] = int(np.mean(label_imu) >= 0.5)
+
+    # Set probability to zero if majority IMU label is 0
+    data_prob[data_label_imu == 0] = 0
+    sqa_label = (data_prob > config.threshold_sqa).astype(np.int8)
+
+    return sqa_label
+
+def extract_hr_segments(sqa_label: np.ndarray, min_hr_samples: int) -> Tuple[np.ndarray, np.ndarray]:
+    """Extracts heart rate segments based on the SQA label.
+
+    Parameters
+    ----------
+    sqa_label : np.ndarray
+        The signal quality assessment label.
+    min_hr_samples : int
+        The minimum number of samples required for a heart rate segment.
+
+    Returns
+    -------
+    Tuple[v_start_idx_long, v_end_idx_long]
+        The start and end indices of the heart rate segments.
+    """
+    # Find the start and end indices of the heart rate segments
+    v_start_idx = np.where(np.diff(sqa_label) == 1)[0] + 1
+    v_end_idx = np.where(np.diff(sqa_label) == -1)[0] + 1
+
+    # Check if the first segment is a start or end
+    if sqa_label[0] == 1:
+        v_start_idx = np.insert(v_start_idx, 0, 0)
+    if sqa_label[-1] == 1:
+        v_end_idx = np.append(v_end_idx, len(sqa_label))
+
+    # Check if the segments are long enough
+    v_start_idx_long = v_start_idx[(v_end_idx - v_start_idx) >= min_hr_samples]
+    v_end_idx_long = v_end_idx[(v_end_idx - v_start_idx) >= min_hr_samples]
+
+    return v_start_idx_long, v_end_idx_long
+
+def extract_hr_from_segment(ppg: np.ndarray, tfd_length: int, fs: int, kern_type: str, kern_params: dict) -> np.ndarray:
+    """Extracts heart rate from the time-frequency distribution of the PPG signal.
+
+    Parameters
+    ----------
+    ppg : np.ndarray
+        The preprocessed PPG segment with 2 seconds of padding on both sides to reduce boundary effects.
+    tfd_length : int
+        Length of each segment (in seconds) to calculate the time-frequency distribution.
+    fs : int
+        The sampling frequency of the PPG signal.
+    MA : int
+        The moving average window length.
+    kern_type : str
+        Type of TFD kernel to use (e.g., 'wvd' for Wigner-Ville distribution).
+    kern_params : dict
+        Parameters for the specified kernel. Not required for 'wvd', but relevant for other 
+        kernels like 'spwvd' or 'swvd'. Default is None.
+
+    Returns
+    -------
+    np.ndarray
+        The estimated heart rate.
+    """
+
+    # Constants to handle boundary effects
+    edge_padding = 4  # Additional 4 seconds (2 seconds on both sides)
```

Comment:
I'd suggest assigning `edge_padding = 4 * fs` and `tfd_length = tfd_length * fs`, since you use `edge_padding` and `tfd_length` only in combination with `fs` in this function, right?

Reply:
agree